docs/contributing/retries-and-waiters.md

```diff
@@ -0,0 +1,474 @@
+		// ... AWS Go SDK logic to update attribute ...
+
+		if err := waiter.ThingAttributeUpdated(conn, d.Id(), d.Get("attribute").(string)); err != nil {
+			return fmt.Errorf("error waiting for Example Thing (%s) attribute update: %w", d.Id(), err)
+		}
+	}
+
+	// ...
+}
+```
+
+## Asynchronous Operations
+
+Many longer running operations are initiated and performed asynchronously in the background after returning a successful response to the initial operation request. These operations may either have a component level status field that can be tracked (e.g. `CREATING`, `UPDATING`, etc.) or they may have an explicit tracking identifier to follow the progress of the background operation.
+
+Terraform resources are expected to wait for these background operations to complete. Failing to do so may introduce incomplete state information or downstream errors in other resources. In rare scenarios involving very long operations, operators may request a flag to skip the waiting, however these should only be implemented case-by-case to prevent those previously mentioned confusing issues.
+
+### AWS Go SDK Waiters
+
+In limited cases, the AWS service API model includes the information to automatically generate a waiter function in the AWS Go SDK for an operation. These are typically named with the prefix `WaitUntil...`. If available, these functions are often good enough for an initial resource implementation and can be used. For example:
+
+```go
+if err := conn.WaitUntilEndpointInService(input); err != nil {
+	return fmt.Errorf("error waiting for Example Thing (%s) ...: %w", d.Id(), err)
+}
+```
+
+If it is necessary to customize the timeouts and polling with these, it is generally preferable to use [Resource Lifecycle Waiters](#resource-lifecycle-waiters) instead, since they are more commonly used throughout the codebase.
+
+### Resource Lifecycle Waiters
+
+Most of the codebase uses `resource.StateChangeConf` and `resource.RefreshStateFunc` handlers for tracking either component level status fields or explicit tracking identifiers. These should be placed in the `aws/internal/service/{SERVICE}/waiter` package and split into separate functions. For example:
+
+```go
+// aws/internal/service/{SERVICE}/waiter/status.go (created if does not exist)
+
+// ThingStatus fetches the Thing and its Status
+func ThingStatus(conn *example.Example, id string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		output, err := /* ... AWS Go SDK operation to fetch resource/status ... */
+
+		if tfawserr.ErrCodeEquals(err, example.ErrCodeResourceNotFoundException) {
+			return nil, "", nil
+		}
+
+		if err != nil {
+			return nil, "", err
+		}
+
+		if output == nil {
+			return nil, "", nil
+		}
+
+		return output, aws.StringValue(output.Status), nil
+	}
+}
+```
+
+```go
+// aws/internal/service/{SERVICE}/waiter/waiter.go (created if does not exist)
+
+const (
+	ThingCreationTimeout = 2 * time.Minute
+	ThingDeletionTimeout = 5 * time.Minute
```

Comment:
In the _Operation Specific Error Retries_ and _Resource Lifecycle Retries_ sections above timeout constants are in `aws/internal/service/{SERVICE}/waiter/consts.go`. Is it confusing to potentially have similar constants in two locations?

Reply:
Fixed in latest commit.